concepts/language-model.md

```diff
@@ -0,0 +1,58 @@
+---
+description: Language model
+---
+
+# Language model
+
+A language model takes text input (usually a history) and outputs the next word or character.
+
+Example:
+- input: `the man is riding a ___`
+- output: `bike` 40%, `car`: 30%, `scooter`: 10%, `blue`: 0.5%, ...
+
+In the example this would mean that the article `bike` has a 40% probability of being the continuation and is the most likely.
+
+It is possible to do *decoding*, which means repeatedly taking e.g. the most probable output, adding it to the text and so on.
+This way a language model can generate new text.
+
+# Models
+
+There are many different ways in which language models are created.
+
+## N-gram maximum likelihood estimate
+
+
+The easiest one is to count a number of occurrences of the pair 
+<img src="https://render.githubusercontent.com/render/math?math=(w_1, w_2)"> and divide that by the number of occurrences of just 
+<img src="https://render.githubusercontent.com/render/math?math=w_1">.
+The result is the probability that with the history (now truncated to just <img src="https://render.githubusercontent.com/render/math?math=w_1">) the next word is <img src="https://render.githubusercontent.com/render/math?math=w_2">.
+Considering only the most recent history is called the Markov assumption.
+
+This model is quite limited because it can't reasonably consider any even mid-term sentence dependencies. 
+This model may be extended to consider longer histories, e.g. 3-grams but there are numerous issues that arise there.
+One solution to those is language model smoothing.
```

Comment:
Maybe we should bold concepts?

Reply:
You mean language model smoothing? That could be a subsection in this article but 

Sentence depdency could be a concept but I wasn't sure because that's becoming quite technical.